miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/js/MiniAppCustomPermissionCache.kt

```diff
@@ -0,0 +1,60 @@
+package com.rakuten.tech.mobile.miniapp.js
+
+import android.content.Context
+import android.content.SharedPreferences
+import androidx.annotation.VisibleForTesting
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+
+/**
+ * A class to read and store the grant results of custom permissions per MiniApp.
+ */
+internal class MiniAppCustomPermissionCache(context: Context) {
+    @VisibleForTesting
+    val prefs: SharedPreferences = context.getSharedPreferences(
+        "com.rakuten.tech.mobile.miniapp.custom.permissions.cache", Context.MODE_PRIVATE
+    )
+
+    /**
+     * Reads the grant results from SharedPreferences.
+     * @param [miniAppId] the key provided to find the stored results per MiniApp
+     * @return [MiniAppCustomPermission] an object to hold the results per MiniApp.
+     */
+    fun readPermissions(miniAppId: String): MiniAppCustomPermission {
+        if (prefs.contains(miniAppId)) {
+            return Gson().fromJson(
+                prefs.getString(miniAppId, ""),
+                object : TypeToken<MiniAppCustomPermission>() {}.type
+            )
+        }
+
+        // if there is no data per miniAppId, then return default values
+        return MiniAppCustomPermission(miniAppId, defaultDeniedList)
+    }
+
+    /**
+     * Stores the grant results to SharedPreferences.
+     * @param [miniAppCustomPermission] an object to hold the results per MiniApp.
+     */
+    fun storePermissions(
+        miniAppCustomPermission: MiniAppCustomPermission
+    ) {
+        val json: String = Gson().toJson(miniAppCustomPermission)
+        prefs.edit().putString(miniAppCustomPermission.miniAppId, json).apply()
```

Comment:
Here it could be possible for the Host App to provide a list of just _some_ of the permissions. For example, they might have already stored `ALLOWED` for `USER_NAME`, `CONTACT_LIST`, and `PROFILE_PHOTO`, but then they call `storePermssions` with something like this:

```
listOf(
    Pair(
        MiniAppCustomPermissionType.USER_NAME,
        MiniAppCustomPermissionResult.DENIED
    ),
    Pair(
        MiniAppCustomPermissionType.CONTACT_LIST,
        MiniAppCustomPermissionResult.DENIED
    )
)
```

In that case, the `PROFILE_PHOTO` permission would be removed and considered denied, but maybe that's not what they intended.

So we could fix this in one of two ways:
1. Change `MiniAppCustomPermission` so that it requires all be permissions be passed. But this might make the interface more difficult to use...
2. OR Change the `storePermissions` method so that it first retrieves the already stored list, and then combines the supplied list with the stored list.

I'm thinking that 2 is probably the better option. What do you think?

Reply:
I will go with option 2.